dropshot/src/websocket.rs

```diff
@@ -0,0 +1,410 @@
+        O: Send + 'static,
+    {
+        // we .take() here to tell Drop::drop that we handled the request.
+        match self.0.take() {
+            None => WebsocketUpgradeResponse {
+                result: Err(HttpError::for_internal_error(
+                    "Tried to handle websocket twice".to_string(),
+                )),
+                join_handle: None,
+            },
+            Some(WebsocketUpgradeInner {
+                upgrade_fut,
+                accept_key,
+                ws_log,
+                ..
+            }) => {
+                let join_handle = tokio::spawn(async move {
+                    match upgrade_fut.await {
+                        Ok(upgrade) => match handler(upgrade).await {
+                            Ok(x) => Ok(x),
+                            Err(e) => {
+                                error!(
+                                    ws_log,
+                                    "Error returned from handler: {:?}", e
+                                );
+                                Err(e)
+                            }
+                        },
+                        Err(e) => {
+                            error!(
+                                ws_log,
+                                "Error upgrading connection: {:?}", e
+                            );
+                            Err(e.into())
+                        }
+                    }
+                });
+                WebsocketUpgradeResponse {
+                    result: Response::builder()
+                        .status(StatusCode::SWITCHING_PROTOCOLS)
+                        .header(header::CONNECTION, "Upgrade")
+                        .header(header::UPGRADE, "websocket")
+                        .header(header::SEC_WEBSOCKET_ACCEPT, accept_key)
+                        .body(Body::empty())
+                        .map_err(Into::into),
+                    join_handle: Some(join_handle),
+                }
+            }
+        }
+    }
+
+    /// Explicitly indicate that we do not wish to upgrade this connection.
+    pub fn do_not_upgrade(mut self) {
+        self.0.take();
+    }
+}
+
+impl Drop for WebsocketUpgrade {
+    fn drop(&mut self) {
+        if let Some(inner) = self.0.take() {
+            panic!("Didn't handle websocket in route {}", inner.route);
+        }
+    }
+}
+
+// To indicate websocket usage by the endpoint to code generators (i.e. Progenitor)
+pub(crate) const WEBSOCKET_EXTENSION: &str = "x-dropshot-websocket";
+pub(crate) const WEBSOCKET_PARAM_SENTINEL: &str = "x-dropshot-websocket-param";
+
+#[derive(JsonSchema)]
+pub struct SchemaWebsocketParams;
+
+impl JsonSchema for WebsocketUpgrade {
+    fn schema_name() -> String {
+        "WebsocketUpgrade".to_string()
+    }
+
+    fn json_schema(
+        gen: &mut schemars::gen::SchemaGenerator,
+    ) -> schemars::schema::Schema {
+        let mut schema = SchemaWebsocketParams::json_schema(gen).into_object();
```

Comment:
I don't think this schema is ever really used in which case I think we could just replace this with
```suggestion
        let mut schema = schemars::schema::SchemaObject::default();
```